src/gt4py/next/new_type_system/traits.py

```diff
@@ -0,0 +1,250 @@
+from __future__ import annotations
+
+import abc
+from typing import Any, Optional
+from collections.abc import Sequence
+
+import dataclasses
+
+
+class Trait:
+    ...
+
+
+class SignednessTrait(Trait):
+    """Specifies the signedness of a type."""
+    @abc.abstractmethod
+    def is_signed(self) -> bool:
+        ...
```

Comment:
Good catch. I added the `ABC` as base class for all of them that had abstract methods.

I don't think this is very well suited for protocols though. Protocols help you decide if an object fits a well-known interface after you've made the object, while abstract base classes help you create an object after you decided it fits a well-known concept. Signedness is a well-known concept, but there is no well-known interface for it in this context, hence the abstract base class that defines the interface.

IMO it's also useful that you can look at a `Type` and tell at a glance what `Trait`s it implements just from the list of base classes instead of having to go through its methods and matching them against all the different `Traits`. There is also the possibility of a `Trait` with no methods, in which case you couldn't use runtime checkable.